### DISCUSSION TITLE
`use()` in a callback or after an `await`?

### DISCUSSION THREAD

**Discussion by busslina** [GregoryConrad/rearch-dart#24] [Q&A] [answered]
I'm pretty sure that using `use` inside a returning callback is a bad practice. But how can I do this?

```dart
void main(List<String> arguments) {
  final capsuleContainer = CapsuleContainer();

  print(capsuleContainer.read(resourceFactoryCapsule)(Resources.datetime));
  print(
      capsuleContainer.read(resourceFactoryCapsule)(Resources.currentUsername));
}

Object Function(String) resourceFactoryCapsule(CapsuleHandle use) {
  return (resourceId) => switch (resourceId) {
        Resources.datetime => DateTime.now(),
        Resources.currentUsername => use(currentUsername),
        _ => throw ('Resouce not implemented'),
      };
}

class Resources {
  static const datetime = 'datetime';
  static const currentUsername = 'current-username';
}

String currentUsername(CapsuleHandle use) => 'Joaquín';
```

https://github.com/busslina/rearch-dart-test/blob/master/lib/main_issue.dart

**Comment by GregoryConrad** [comment, marked as answer]
Yup, it is a bad practice because it causes the dependency graph to not be known at the conclusion of build time. If you're on the latest version of ReArch (released a day or two ago), you should be getting an `AssertionError` when you actually invoke the callback (if you're up-to-date and not getting an error, please let me know as that would be a bug).

See [here](https://rearch.gsconrad.com/paradigms/factories#an-example) for how to correctly do this.

## The Problem

While unlikely to cause _significant_ issues in most situations, it can break the consistency guarantee of containers in certain downstream side effects. Take the following:

```dart
Future<bool> myAsyncCapsule(CapsuleHandle use) {
  await use(someOtherFuture); // safe
  return use(someBool); // unsafe, after await
}

void loggerCapsule(CapsuleHandle use) {
  use.asListener();
  // This is a rather useless example, but you can get the picture:
  print(use(myAsyncCapsule));
}
```

If `someBool` changes while `someOtherFuture` is completing, `myAsyncCapsule` will not be rebuilt because the container would have no idea that `myAsyncCapsule` depends upon `someBool` until it gets `use`d. Thus, if you have `loggerCapsule` going (that reads the output of this `myAsyncCapsule`), it will only get the state update of `myAsyncCapsule` once (with the new value), instead of twice, as you would expect. A bigger issue is a scenario like this:

```dart
Future<bool> anotherAsyncCapsule(CapsuleHandle use) async {
  if (use(aCapsule)) await use(someAsyncCapsule);
  return use(aCapsule);
}
```

Now imagine here that `aCapsule` mutates while we are awaiting `someAsyncCapsule`--whoops! That is definitely not what we were intending for `anotherAsyncCapsule` to return, and that again breaks consistency guarantees of a container.

**Comment by GregoryConrad** [comment]
Also, returning `Object` tends to be a code smell--unless you are just trying to `toString()` or something on it. You often should use a type map or return a common interface, depending on what you need.

**Reply by busslina** [reply]
I want to replicate a resource managing and synchronizing module (backend <-> frontend) that I have fully functional with Riverpod.
I use `Object` because a resource can be anything (I determine its type by its resource key, and it works great). I know how to do factory capsules. The problem I have here is to be able to depend on an undetermined capsule, only determined by the `String` key. In this case I have the key in the returned callback.

Maybe is not possible by design. Riverpod allows you to depend on other providers determined by the key (conditionally).

Don't you have any solution for this issue? Maybe using more capsules?
I keep thinking about a solution.

**Reply by busslina** [reply]
Maybe a solution could be to have the correct `CapsuleHandle` as a paramter on the returned callback. What do you think about this?

I mean, not the `CapsuleHandle` itself, but a function that returns the correct `CapsuleHandle`, so I can use it in the callback function.

**Reply by GregoryConrad** [reply]
Could you please specify the exact behavior you want to achieve? I.e., the exact requirements and input/output of this synchronization component. Or, I can take a look at what you currently do in Riverpod to give you an analogous solution.

Sounds like you may need inline capsules (closures that are capsules because they consume a `CapsuleHandle`), but I'd need to know exactly what you need to accomplish first before I steer you in a particular direction.

**Reply by busslina** [reply]
Here you have the two main classes, one for backend and the other for frontend:
https://gist.github.com/busslina/4c9684fa9fea3512397d5a559cc36373

Here you have the class `ResourceIdentifier`:
```dart
@freezed
class ResourceIdentifier with _$ResourceIdentifier, llib.Serializable {
  const ResourceIdentifier._();

  const factory ResourceIdentifier({
    required String key,
    String? extraKey,
  }) = _ResourceIdentifier;

  factory ResourceIdentifier.fromJson(JsonMap json) =>
      _$ResourceIdentifierFromJson(json);

  void debug(String msg) => llib.debug('$this -- $msg');
}
```

And here you have the current resources:
```dart
class ResourceKeys {
  static const hostname = 'hostname';
  static const defaultUser = 'default-user';
  static const datetime = 'datetime';
  static const timezone = 'timezone';
  static const timezones = 'timezones';
  static const bashrcPatchedRootUser = 'bashrc-patched-root-user';
  static const bashrcPatchedDefaultUser = 'bashrc-patched-default-user';
  static const osUserList = 'os-user-list';
  static const osUser = 'os-user';
}

class ResourceIds {
  static const hostname = ResourceIdentifier(
    key: ResourceKeys.hostname,
  );

  static const defaultUser = ResourceIdentifier(
    key: ResourceKeys.defaultUser,
  );

  static const datetime = ResourceIdentifier(
    key: ResourceKeys.datetime,
  );

  static const timezone = ResourceIdentifier(
    key: ResourceKeys.timezone,
  );

  static const timezones = ResourceIdentifier(
    key: ResourceKeys.timezones,
  );

  static const bashrcPatchedRootUser = ResourceIdentifier(
    key: ResourceKeys.bashrcPatchedRootUser,
  );

  static const bashrcPatchedDefaultUser = ResourceIdentifier(
    key: ResourceKeys.bashrcPatchedDefaultUser,
  );

  static const osUserList = ResourceIdentifier(
    key: ResourceKeys.osUserList,
  );

  static ResourceIdentifier osUser(String userId) => ResourceIdentifier(
        key: ResourceKeys.osUser,
        extraKey: userId,
      );
}
```


There are some other code needed to make it work, like websocket handlers, utility extensions and other providers like the one who store the resource version for a resource (IntegrityIndex) but the idea is pretty simple, the backend provider can hold any type of resource and manages its value/versions across multiple frontends.

**Reply by GregoryConrad** [reply]
Would you mind explaining why are you bunching all of these separate resources (with separate types) together into one provider just so I can get a better idea of what you are trying to do here? Are you sharing some central logic between them all?

Here are my initial thoughts:
You often should favor the composition of smaller components versus just having one big component, which it seems like you're currently doing (this is true in Riverpod too). I.e., make a capsule for each of your resources individually, and if they share some common functionality, you can make a shared side effect to handle that common ground.

**Reply by GregoryConrad** [reply]
Also, sometimes that common functionality may not just be a side effect, but rather a shared dependency capsule. Depends on the scenario (shared state across your resources -> shared dependency capsule. each capsule has it's own state, but that state logic is the same -> shared side effect)

**Reply by busslina** [reply]
I put it together for multiple reasons.
`Resources` have this features:
- They are catched.
- They are versioned.
- They are shared with clients.
- They are changed listened.
- They are broadcasted on changed to listeners.
- They are locked prior edition.

I share with you the remaining providers (only the backend ones) associated to better understand the whole feature that are `Resources`. Imagine replicate all that features for every `Resource`. Would be crazy... But I'm opened to `rearch` change my structure and be able to use composition. That yes, but not to replicate code over and over.
Anyway, the problem that came up today will be always present. (I think)

Resource lock provider:
```dart
@Riverpod(keepAlive: false)
class ResourceLock extends _$ResourceLock {
  KeepAliveLink? _keepAliveLink;
  ProviderSubscription<commons.ConnectionState>? _serverConnectionSubscription;

  CancelableOperation? _releaseTimeoutOperation;

  /// Initializes the provider with non acquired state. It will autodispose if
  /// doesn't gets acquired.
  @override
  ResourceLockState build(ResourceIdentifier resource) {
    ref.debugStateChanges(
      'resourceLockProvider -- $resource',
      active: current.Instances.debug.resourceLockProvider,
    );

    ref.onDispose(() => _releaseTimeoutOperation?.cancel());

    return ResourceLockState(
      resource: resource,
      holderConnectionKey: null,
    );
  }

  /// Tries to acquire the lock. If it's already acquired, returns true if the holder
  /// is the current connection, or false if it isn't.
  /// When acquired, the provider becomes keept alive.
  /// The lock will be released after [releaseTimeout] if it's set.
  bool acquire(
    ConnectionKey holderConnectionKey, {
    Duration releaseTimeout = Duration.zero,
  }) {
    if (state.active) {
      return state.holderConnectionKey == holderConnectionKey;
    }

    _incrementIntegrityIndex();

    state = state.copyWith(holderConnectionKey: holderConnectionKey);

    _keepAliveLink = ref.keepAlive();

    // Release timeout operation
    if (releaseTimeout != Duration.zero) {
      (_releaseTimeoutOperation =
              CancelableOperation.fromFuture(delay(releaseTimeout)))
          .value
          .then((_) => release());
    }

    // Listening to connection state changes
    _serverConnectionSubscription = ref.listenConnectionStateChanges(
        holderConnectionKey, () => release(adviceClient: false));

    _adviceClient(holderConnectionKey);

    return true;
  }

  /// Tries to release the lock. If it's not acquired or it's not the current
  /// holder, returns false.
  ///
  /// When released, the provider becomes autodispose.
  bool release({
    ConnectionKey? connectionKey,
    bool adviceClient = true,
  }) {
    if (!state.active ||
        (connectionKey != null && !state.checkOwned(connectionKey))) {
      return false;
    }

    _incrementIntegrityIndex();

    final holderConnectionKey = state.holderConnectionKey!;

    state = state.copyWith(holderConnectionKey: null);

    _keepAliveLink?.close();

    _releaseTimeoutOperation?.cancel();

    _serverConnectionSubscription?.close();

    if (adviceClient) {
      _adviceClient(holderConnectionKey);
    }

    return true;
  }

  void _adviceClient(ConnectionKey holderConnectionKey) => ref
      .read(commons.connectionProvider(holderConnectionKey).notifier)
      .sendMessageResource(commons.ResourceMessageData(
        event: state.active
            ? commons.ResourceSignals.lockAcquired
            : commons.ResourceSignals.lockRevoked,
        resource: resource,
        data: _integrityIndex,
      ));

  int get _integrityIndex => ref.read(integrityIndexProvider(
        IntegrityIndexType.resourceLock,
        extraKey: resource,
      ));

  int _incrementIntegrityIndex() => ref
      .read(integrityIndexProvider(
        IntegrityIndexType.resourceLock,
        extraKey: resource,
      ).notifier)
      .increment();
}

@freezed
class ResourceLockState with _$ResourceLockState {
  const ResourceLockState._();

  const factory ResourceLockState({
    required ResourceIdentifier resource,
    ConnectionKey? holderConnectionKey,
  }) = _ResourceLockState;

  bool get active => holderConnectionKey != null;

  bool checkOwned(Object connectionKey) => holderConnectionKey == connectionKey;
}
```

Integrity index:
```dart
@Riverpod(keepAlive: true)
class IntegrityIndex extends _$IntegrityIndex {
  @override
  int build(IntegrityIndexType type, {Object? extraKey}) => 0;

  int increment() => ++state;
}

enum IntegrityIndexType {
  pluginState,
  pluginAuthenticatedState,
  resourceLock,
  resource,
}
```

Resource listeners:
```dart
/// Holds a set of connection keys which are listening to the resource.
@Riverpod(keepAlive: true)
class ResourceListener extends _$ResourceListener {
  static const _debugging = true;

  final _subscriptions =
      HashMap<ConnectionKey, ProviderSubscription<commons.ConnectionState>>();

  @override
  ISet<ConnectionKey> build(ResourceIdentifier resource) => ISet();

  void addListener(ConnectionKey connectionKey) {
    state = state.add(connectionKey);

    // Listening to connection state changes (if didn't yet)
    _subscriptions[connectionKey] ??=
        ref.listenConnectionStateChanges(connectionKey, () {
      final removed =
          removeListener(connectionKey, debugTag: 'connection state changed');

      debug(
        'ResourceListener -- Connection state changed -- Removed listener: $removed -- $resource',
        active: _debugging,
      );
    });

    debug(
      'ResourceListener.addListener() -- $resource -- $state',
      active: _debugging,
    );
  }

  bool removeListener(ConnectionKey connectionKey, {required String debugTag}) {
    state = state.remove(connectionKey);

    _subscriptions[connectionKey]?.close();
    final removed = _subscriptions.remove(connectionKey);

    debug(
      'ResourceListener.removeListener() -- $debugTag -- $resource -- $state',
      active: _debugging,
    );

    return removed != null;
  }
}
```

Resource editable:
```dart
@Riverpod(keepAlive: true)
class ResourceEditable extends _$ResourceEditable {
  @override
  bool build(ResourceIdentifier resource) => _isEditable();

  bool _isEditable() => switch (resource) {
        commons.ResourceIds.hostname => true,
        commons.ResourceIds.defaultUser => true,
        commons.ResourceIds.bashrcPatchedRootUser => true,
        commons.ResourceIds.bashrcPatchedDefaultUser => true,
        _ => false,
      };
}
```

**Reply by busslina** [reply]
I want to move that step by step to `rearch` but it always came the problem. To use `CapsuleHandle` once you have the `family key`

**Reply by busslina** [reply]
Sorry, for attacking you with tones of code haha.

I'm just thinking before going to sleep that maybe I could avoid that problem by do a `switch` in the WS handler that redirects to the correct capsule. There I will not need any key because I'm in the correct capsule and I can redo all this logic.

I will dream about this and tomorrow I will start this way.

What do you think it about it? Feasible?

**Reply by GregoryConrad** [reply]
This is starting to look like an XY problem--you don't actually want factories/families here, you want individual capsules that all have some common ground.

> I put it together for multiple reasons.
> Resources have this features:
> - They are catched.
> - They are versioned.
> - They are shared with clients.
> - They are changed listened.
> - They are broadcasted on changed to listeners.

Ok! So that's something we can work with.

> They are catched

Assuming you meant "cached": No problem, make each resource its own capsule. Capsules are cached automatically.

> They are versioned

If all resources are combined into entire "app" versions that act as state snapshots, you can use a singular global versioning capsule to handle this. Otherwise, a side effect to handle each capsule's versioning will work well.

> They are shared with clients

One capsule per resource would work well here.

> They are changed listened.
> They are broadcasted on changed to listeners.

What are they listened to by? Other capsules, some generic listener, or open client socket connections? Or all of the above?
For other capsules or listeners, no worries--changes automatically propagate.
For open client connections, create a side effect per resource capsule that handles a set of listeners.

A pseudo-example:

```dart
({
  Datetime date,
  void Function() setDate(Datetime),
  int version,
  // other things each resource might have in common...
}) datetimeResourceCapsule(CapsuleHandle use) {
  final (version, incrementVersion) = use.resourceVersion();
  return (
    date: OsUtils.whateverTheFunctionIs(),
    setDate: (newDate) {
      OsUtils.whateverTheFunctionIs(newDate);
      incrementVersion();
    },
    version: version,
  );
}
```

Each resource can expose a similar sort of capsule API. If you need to be able to access a list of resource versions or similar, you can make some more capsules:

```dart
int datetimeResourceVersionCapsule(CapsuleHandle use) {
  return use(dateTimeResourceCapsule).version;
}

int Function(ResourceId) resourceVersionsAction(CapsuleHandle use) {
  return (resource) => switch (resource) { ... };
}
```

**Reply by GregoryConrad** [reply]
> I'm just thinking before going to sleep that maybe I could avoid that problem by do a switch in the WS handler that redirects to the correct capsule. There I will not need any key because I'm in the correct capsule and I can redo all this logic.

Not sure exactly what you are doing currently, but that does sound like a reasonable idea--client requests can then be routed to the applicable capsule.

**Reply by busslina** [reply]
Thanks! I will let you know if I have success with this.

**Reply by busslina** [reply]
I meant listened by WS connections AKA by clients

**Reply by GregoryConrad** [reply]
Ok, then something like this would work well:

```dart
({void Function(Connection) addLIstener, void Function(Connection) removeListener}) datetimeResourceListener(CapsuleHandle use) {
  final set = use.value(<Connection>{});
  for (final listener in set) listener.send(use(datetimeResourceCapsule).datetime);
  return (
    addListener: set.add,
    removeListener: set.remove,
  );
}
```

**Reply by busslina** [reply]
Yeah, I was thinking something like this just now. But having all together in the same capsule. Resource value, listeners, version number, lock, ...

What do you think? All in one capsule or better in separated capsules?

There's here a new question. Can I avoid replicate all this code for every resource? Maybe a new registered side effect?

Then I would have all capsules equally signed and using a typedef, so I can call a function that returns the capsule associated to a resource. Then automate the capsule handling from the WS handler

**Reply by busslina** [reply]
And why are you sending to listeners there?
Isn't better to call it inside an update capsule value function?

**Reply by GregoryConrad** [reply]
> What do you think? All in one capsule or better in separated capsules?

I’d probably make one main capsule per resource, with one additional capsule for each resource to handle the listeners. The listeners might be more ergonomic/easier to handle in a separate capsule.

You *might* actually be able to combine the listener capsules into the main capsules if you really wanted via a reusable side effect, but this would come with the (slight) disadvantage that you need to manually pass the current data to this side effect instead of it automatically getting it.

> There's here a new question. Can I avoid replicate all this code for every resource? Maybe a new registered side effect?

This is exactly what side effect composability is for!! You could do something like the following if you wanted:
```dart
ResourceObject<DateTime> datetimeResourceCapsule(CapsuleHandle use) {
  return use.asResource<Datetime>(
    getResource: OsUtils.dateTime,
    setResource: OsUtils.setDateTime,
    // other stuff your side effect may need
  );
}

// and resource object can contain methods to add/remove listeners, get a lock etc.
// the side effect would also automatically send resource updates to the listeners
// (note how the object has a way to get the resource--thus, it can get it and send it to listeners as needed)
```

> Then I would have all capsules equally signed and using a typedef, so I can call a function that returns the capsule associated to a resource. Then automate the capsule handling from the WS handler

That doesn't seem quite right--id still probably manually write one capsule out per resource, like I did with `datetimeResourceCpasule` above.

> And why are you sending to listeners there?
Isn't better to call it inside an update capsule value function?

The separate capsule just handles a set of listeners, and whenever the resource capsule updates, the listener capsule will get invoked again and thus be able to send all of the current listeners an update. It's a bit more declarative and easier to maintain than the imperative approach where you manually invoke each listener, but if you are making one "mega" resource side effect, having it integrated in the same side effect would be acceptable.

**Reply by busslina** [reply]
Hi again,

what do you think about this?

#### 1) resource.side_effects.dart
```dart
extension _UseConvenience on SideEffectRegistrar {
  SideEffectRegistrar get use => this;
}

extension ResourceSideEffects on SideEffectRegistrar {
  ResourceCapsuleSideEffectResult<T> asResource<T extends Object>({
    required FutureOr<T?> initialValue,
  }) =>
      register((api) {
        final (value, setValue) =
            use.state(VersionedFutureOrValue.fresh(initialValue));

        final listeners = use.value(<ConnectionKey>{});

        final (lock, setLock) =
            use.state(VersionedValue<ConnectionKey>.fresh(null));

        return (
          // (01) Value
          valueManager: (
            value: value.value,
            setValue: (newValue) {
              if (value.value == newValue) {
                return false;
              }

              setValue(VersionedFutureOrValue(newValue, value.version + 1));

              // TODO: execute listeners: value changed

              api.rebuild();

              return true;
            },
          ),

          // (02) Version
          version: value.version,

          // (03) Listeners
          listenersManager: (
            add: (connectionKey) {
              final res = listeners.add(connectionKey);

              if (!res) {
                return false;
              }

              // TODO: listen to connection state change and remove listener

              // TODO: advice frontend: listener on

              return true;
            },
            remove: (connectionKey) {
              final res = listeners.remove(connectionKey);

              if (!res) {
                return false;
              }

              // TODO: cancel and remove subscription to connection state change

              // TODO: advice frontend: listener off
              // (not if release is triggered by connection state change)

              return true;
            },
          ),

          // (04) Lock
          lockManager: (
            lock: lock.value,
            version: lock.version,
            acquire: (connectionKey) {
              if (lock.value == connectionKey) {
                return true;
              }

              if (lock.value != null) {
                return false;
              }

              setLock(VersionedValue(connectionKey, lock.version + 1));

              // TODO: add support for auto-release on timeout

              // TODO: listen to connection state change and release lock

              // TODO: advice frontend: lock acquired

              return true;
            },
            release: (connectionKey) {
              if (lock.value == null) {
                return false;
              }

              if (lock.value != connectionKey) {
                return false;
              }

              setLock(VersionedValue(null, lock.version + 1));

              // TODO: cancel and remove subscription to auto-release on timeout

              // TODO: cancel and remove subscription to connection state change

              // TODO: advice frontend: lock released
              // (not if release is triggered by connection state change)

              return true;
            },
          ),
        );
      });
}

typedef ResourceCapsuleSideEffectResult<T extends Object> = ({
  _ValueManager<T> valueManager,
  int version,
  _ListenersManager listenersManager,
  _LockManager lockManager,
});

typedef _ValueManager<T extends Object> = ({
  FutureOr<T?> value,
  bool Function(FutureOr<T?>) setValue,
});

typedef _ListenersManager = ({
  bool Function(ConnectionKey) add,
  bool Function(ConnectionKey) remove,
});

typedef _LockManager = ({
  ConnectionKey? lock,
  int version,
  bool Function(ConnectionKey) acquire,
  bool Function(ConnectionKey) release,
});

class VersionedValue<T extends Object> with EquatableMixin {
  // final FutureOr<T?> value;
  final T? value;
  final int version;

  VersionedValue(this.value, this.version);

  VersionedValue.fresh(this.value) : version = 0;

  @override
  List<Object?> get props => [value, version];
}

class VersionedFutureOrValue<T extends Object> with EquatableMixin {
  final FutureOr<T?> value;
  final int version;

  VersionedFutureOrValue(this.value, this.version);

  VersionedFutureOrValue.fresh(this.value) : version = 0;

  @override
  List<Object?> get props => [value, version];
}
```

#### 2) datetime_resource.capsule.dart
```dart
Future<DatetimeResource> datetimeResource(CapsuleHandle use) async {
  final resourceManager = use.asResource<DateTime>(initialValue: now());

  return (
    value: (await resourceManager.valueManager.value)!,
    version: resourceManager.version,
    isEditable: true,
    update: (newValue) => resourceManager.valueManager.setValue(newValue),
    listeners: resourceManager.listenersManager,
    lock: resourceManager.lockManager,
  );
}

typedef DatetimeResource = ({
  DateTime value,
  int version,
  bool isEditable,
  void Function(DateTime) update,
  _ListenersManager listeners,
  _LockManager lock,
});

typedef _ListenersManager = ({
  bool Function(ConnectionKey) add,
  bool Function(ConnectionKey) remove,
});

typedef _LockManager = ({
  ConnectionKey? lock,
  int version,
  bool Function(ConnectionKey) acquire,
  bool Function(ConnectionKey) release,
});
```

Should I call `api.rebuild()` only on `setValue` or should I also call it in other parts like "lock acquire" or "lock release"?

**Reply by busslina** [reply]
BTW it produces much less code than Riverpod and more elegant.

**Reply by GregoryConrad** [reply]
Wow, that is looking so much better! Easy to follow.

> Should I call api.rebuild() only on setValue or should I also call it in other parts like "lock acquire" or "lock release"?

that was my only comment: you don’t need to call rebuild because setValue will already call it for you! State setters internally call rebuild.

**Reply by busslina** [reply]
> that was my only comment: you don’t need to call rebuild because setValue will already call it for you! State setters internally call rebuild.

I was going to create a new question... but you answered here before I ask hehe.

I was mastering the `countdown` test and I was getting angry because the widget was updating before the timer callback executing (the rebuild). It's caused by `setState`:
```dart
// setTimer(startTimer());
        startTimer();
```

Related questions:
1) Only `setState` trigger rebuilds internally?
2) Do `rebuild()` is triggered immediately? Or I can call it multiple times in a row but only one re-build will execute?
3) So the only purpose of `register` is to provide the api? If so, seems that putting code inside its parameter callback is useless and in some cases dangerous. Wouldn't be better to only have an api getter instead (with no callback)?

**Reply by GregoryConrad** [reply]
Those are really good questions; I should add them to the FAQ! The fact that you’re asking these means you’re starting to pick it up pretty well 😉

> Only setState trigger rebuilds internally?

No, any of the side effects that mutate some data and then return some form of that data to the capsule/widget will trigger rebuilds.

> Do rebuild() trigger state immediately? Or I can call it multiple times in a row but only one re-build will execute?

for capsules: yes. Calling multiple times in a row will trigger multiple rebuilds. Since that is sometimes undesirable. We can make “side effect transactions” with use.transactionRunner() that can batch up any number of calls to rebuild and handle them all in one go.

for widgets: no, rebuilds are batched together automatically due to flutter’s build scheduler (ReArch uses Element.markNeedsBuild or whatever it is called to rebuild widgets)

> So the only purpose of register is to provide the api? If so, seems that putting code inside its parameter callback is useless and in some cases dangerous. Wouldn't be better to only have an api getter instead (with no callback)?

it’s to provide the API and to run some code _only on the first build_. I suppose I could’ve instead implemented the SideEffectRegistrar in Dart as:
- SideEffectApi get api;
- T callonce(T Function())

but i didn’t, and now that’d be a breaking change now 😝

If for some odd reason I do need to ever release a v2.0, that might be a change to consider to make the lower-level side effects more understandable.

But to address the concern you raised: code inside register() is not useless, as it is only ever called on the first build and the resulting data is returned on every subsequent build; this lets you initialize some state or hold the same state across builds.

**Reply by busslina** [reply]
Thank you ;)

> for capsules: yes. Calling multiple times in a row will trigger multiple rebuilds. Since that is sometimes undesirable. We can make “side effect transactions” with use.transactionRunner() that can batch up any number of calls to rebuild and handle them all in one go.

I need to study transactions.

> No, any of the side effects that mutate some data and then return some form of that data to the capsule/widget will trigger rebuilds.

Do you mean these ones? Am I missing any?
- lazyStateGetterSetter
- stateGetterSetter
- lazyState
- state
- memo (I think maybe not)

When I say elegant is for example this code:
```dart

T? previous<T>(T current) {
  final (getter, setter) = use.rawValueWrapper<T?>(() => null);
  final prev = getter();
  setter(current);
  return prev;
}

bool isFirstBuild() => use.previous(false) ?? true;
```

> But to address the concern you raised: code inside register() is not useless, as it is only ever called on the first build and the resulting data is returned on every subsequent build; this lets you initialize some state or hold the same state across builds.

Maybe you should specify this in the code doc because the current doc for `register` is `Registers the given side effect.`.
Now I understand why `callonce` and others call `register`.

**Reply by busslina** [reply]
You should be doing Flutter YouTube interviews explaining this Lib. (I mean for the official Flutter/Google channel)

**Reply by busslina** [reply]
> Do you mean these ones? Am I missing any?
lazyStateGetterSetter
stateGetterSetter
lazyState
state
memo (I think maybe not)

Okay, I see. Must be all dependants on `lazyStateGetterSetter`.

**Reply by GregoryConrad** [reply]
> I need to study transactions

not much to study thankfully! Just do:

`runTransaction(() { yourSideEffectsThatRebuildHere(); });`

> Maybe you should specify this in the code doc because the current doc for register is Registers the given side effect..
Now I understand why callonce and others call register.

got this update to the docs locally, should push it soon. Thanks for the suggestion!

> Do you mean these ones? Am I missing any?

anything that involves mutating some internal side effect state (not within the build itself) and has that state returned from the side effect in some capacity will likely trigger a rebuild. Thus, all of the state side effects, reducer, persist, mutation, and a couple others will trigger rebuilds in their callbacks.
Memo will not trigger a rebuild, because while it does have some state that is mutated, that state is only ever mutated from within build so there is no need to ever rebuild (because you’re already in build!).

> Must be all dependants on lazyStateGetterSetter.

while that is true for many of the side effects, some others will trigger rebuilds like invalidatableFuture (iirc). Just try to follow that rule of thumb from above for how the data might mutate and where.

> You should be doing Flutter YouTube interviews explaining this Lib. (I mean for the official Flutter/Google channel)

well, if ReArch ever got popular enough, I’d love to do it!

> Is this (extension) provided or I must create it manually?

it’s provided out of the box! See https://pub.dev/documentation/rearch/latest/rearch/BuiltinSideEffects.html

**Reply by GregoryConrad** [reply]
Whoops, wrong link. See here: https://pub.dev/documentation/rearch/latest/rearch/CapsuleMapper.html
